site/src/pages/RoadmapPage/ImportTranscriptPopup.tsx

```diff
@@ -0,0 +1,243 @@
+  const text = row.text.trim();
+  if (!text) return;
+
+  if (row.classList.contains('title')) {
+    quarters.push({ name: text, courses: [] });
+    return;
+  }
+
+  // No quarters at UCI Yet; these are transferred credits
+  if (!quarters.length) {
+    if (text.startsWith('Units Transferred')) return;
+    const [raw, name, score, units, date] =
+      text.match(/^((?:.(?!\(Score|\(Units))+)\s\((?:Score\s(\d))?(?:,\s)?Units ([\d.]+)\).*?(\d+\/\d+)/) || [];
+    if (!raw) return;
+    transfers.push({ raw, name, score: Number(score), units: Number(units), date });
+    return;
+  }
+
+  if (!row.classList.contains('grades')) return;
+
+  // This is a course taken at UCI, add it to the last seen quarter title
+  const cols = row.childNodes.filter((x) => x.nodeType === 1).slice(1);
+  const [title, dept, code, units, grade, gradePts] = cols.map((x) => x.text.trim());
+
+  // Add this course to the current (last) item in the quarters array
+  quarters.at(-1)!.courses.push({ title, dept, code, units: Number(units), grade, gradePts: Number(gradePts) });
+}
+
+async function htmlFromFile(file: Blob): Promise<HTMLElement> {
+  const reader = new FileReader();
+  const fileText = (await new Promise((resolve) => {
+    reader.addEventListener('load', () => resolve(reader.result as string));
+    reader.readAsText(file);
+  })) as string;
+  return parseHTML(fileText);
+}
+
+type CourseLookup = { [k: string]: CourseGQLData };
+
+async function transcriptCourseDetails(quarters: TranscriptQuarter[]): Promise<CourseLookup> {
+  const courseEntries = quarters
+    .flatMap((q) => q.courses)
+    .map((c, i) => {
+      const dept = JSON.stringify(c.dept);
+      const cNum = JSON.stringify(c.code);
+      return `_${i}: courses(department: ${dept}, courseNumber: ${cNum}) { ...Details }`;
+    });
+
+  const query = `fragment Details on Course {
+    id, department, courseNumber, school, title, courseLevel,
+    minUnits, maxUnits, description, departmentName,
+    instructors { ucinetid, name, shortenedName },
+    prerequisiteTree, prerequisiteText,
+    prerequisites { id, department, courseNumber, title },
+    dependencies { id, department, courseNumber, title },
+    repeatability, concurrent, sameAs, restriction, overlap,
+    corequisites, geList, geText, terms
+  }
+  query CourseDetails { ${courseEntries.join(',\n')} }`;
+
+  const response = await fetch('https://api-next.peterportal.org/v1/graphql', {
+    method: 'POST',
+    headers: { 'content-type': 'application/json' },
+    body: JSON.stringify({ query }),
+  }).then((x) => x.json());
+
+  // Query Response is { data: { [key: string]: [CourseData] } }
+  // Transform this response into a lookup { courseID: {CourseData} }
+  return Object.fromEntries(
+    Object.values<CourseGQLData[]>(response.data)
+      .flat()
+      .map((c) => [c.department + c.courseNumber, c]),
+  );
+}
```

Comment:
use `searchAPIResults` helper function instead

Reply:
It's not a lookup by course ID, it's a lookup by department and code. It seems I could get the ID by simply removing all the spaces, but that seems like a fragile solution, even if that is consistent for every entry in the API.

I do think it would be good to not have this duplicate code though – do you have any suggestions for the best way to fix this?

Reply:
Replacing spaces with empty strings seems fine to me. That's how the API's scrapers generate the course id.

Reply:
> that seems like a fragile solution, even if that is consistent for every entry in the API

To be honest I still stand by this but I think this is the best we can do for now.